File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloCacheInterceptor.java
package com.apollographql.apollo.internal.interceptor;

import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.internal.Function;
import com.apollographql.apollo.api.internal.Optional;
import com.apollographql.apollo.cache.normalized.ApolloStore;
import com.apollographql.apollo.cache.normalized.ApolloStoreOperation;
import com.apollographql.apollo.cache.normalized.Record;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.interceptor.ApolloInterceptor;
import com.apollographql.apollo.interceptor.ApolloInterceptorChain;
import com.apollographql.apollo.internal.cache.normalized.ResponseNormalizer;
import com.apollographql.apollo.internal.cache.normalized.Transaction;
import com.apollographql.apollo.internal.cache.normalized.WriteableStore;
import com.apollographql.apollo.internal.util.ApolloLogger;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Executor;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

/**
 * ApolloCacheInterceptor is a concrete {@link ApolloInterceptor} responsible for serving requests from the normalized
 * cache if {@link InterceptorRequest#fetchFromCache} is true. Saves all network responses to cache.
 */
public final class ApolloCacheInterceptor implements ApolloInterceptor {
  private final ApolloStore apolloStore;
  private final ResponseFieldMapper responseFieldMapper;
  private final Executor dispatcher;
  private final ApolloLogger logger;
  private volatile boolean disposed;

  public ApolloCacheInterceptor(@Nonnull ApolloStore apolloStore, @Nonnull ResponseFieldMapper responseFieldMapper,
      @Nonnull Executor dispatcher, @Nonnull ApolloLogger logger) {
    this.apolloStore = checkNotNull(apolloStore, "cache == null");
    this.responseFieldMapper = checkNotNull(responseFieldMapper, "responseFieldMapper == null");
    this.dispatcher = checkNotNull(dispatcher, "dispatcher == null");
    this.logger = checkNotNull(logger, "logger == null");
  }

  @Override
  public void interceptAsync(@Nonnull final InterceptorRequest request, @Nonnull final ApolloInterceptorChain chain,
      @Nonnull final Executor dispatcher, @Nonnull final CallBack callBack) {
    dispatcher.execute(new Runnable() {
      @Override public void run() {
        if (disposed) return;
        if (request.fetchFromCache) {
          callBack.onFetch(FetchSourceType.CACHE);
          final InterceptorResponse cachedResponse;
          try {
            cachedResponse = resolveFromCache(request);
            callBack.onResponse(cachedResponse);
            callBack.onCompleted();
          } catch (ApolloException e) {
            callBack.onFailure(e);
          }
        } else {
          writeOptimisticUpdatesAndPublish(request);
          chain.proceedAsync(request, dispatcher, new CallBack() {
            @Override public void onResponse(@Nonnull InterceptorResponse networkResponse) {
              if (disposed) return;

              try {
                Set<String> networkResponseCacheKeys = cacheResponse(networkResponse, request);
                Set<String> rolledBackCacheKeys = rollbackOptimisticUpdates(request);
                Set<String> changedCacheKeys = new HashSet<>();
                changedCacheKeys.addAll(rolledBackCacheKeys);
                changedCacheKeys.addAll(networkResponseCacheKeys);
                publishCacheKeys(changedCacheKeys);
              } catch (Exception rethrow) {
                rollbackOptimisticUpdatesAndPublish(request);
                throw rethrow;
              }

              callBack.onResponse(networkResponse);
              callBack.onCompleted();
            }

            @Override public void onFailure(@Nonnull ApolloException t) {
              rollbackOptimisticUpdatesAndPublish(request);
              callBack.onFailure(t);
            }

            @Override public void onCompleted() {
            }

            @Override public void onFetch(FetchSourceType sourceType) {
              callBack.onFetch(sourceType);
            }
          });
        }
      }
    });
  }

  @Override public void dispose() {
    disposed = true;
  }

  private InterceptorResponse resolveFromCache(InterceptorRequest request) throws ApolloException {
    ResponseNormalizer<Record> responseNormalizer = apolloStore.cacheResponseNormalizer();
    //noinspection unchecked
    ApolloStoreOperation<Response> apolloStoreOperation = apolloStore.read(request.operation, responseFieldMapper,
        responseNormalizer, request.cacheHeaders);
    Response cachedResponse = apolloStoreOperation.execute();
    if (cachedResponse.data() != null) {
      logger.d("Cache HIT for operation %s", request.operation);
      return new InterceptorResponse(null, cachedResponse, responseNormalizer.records());
    }
    logger.d("Cache MISS for operation %s", request.operation);
    throw new ApolloException(String.format("Cache miss for operation %s", request.operation));
  }

  private Set<String> cacheResponse(final InterceptorResponse networkResponse,
      final InterceptorRequest request) {
    final Optional<List<Record>> records = networkResponse.cacheRecords.map(
        new Function<Collection<Record>, List<Record>>() {
          @Nonnull @Override public List<Record> apply(@Nonnull Collection<Record> records) {
            final List<Record> result = new ArrayList<>(records.size());
            for (Record record : records) {
              result.add(record.toBuilder().mutationId(request.uniqueId).build());
            }
            return result;
          }
        }
    );

    if (!records.isPresent()) {
      return Collections.emptySet();
    }

    try {
      return apolloStore.writeTransaction(new Transaction<WriteableStore, Set<String>>() {
        @Nullable @Override public Set<String> execute(WriteableStore cache) {
          return cache.merge(records.get(), request.cacheHeaders);
        }
      });
    } catch (Exception e) {
      logger.e("Failed to cache operation response", e);
      return Collections.emptySet();
    }
  }

  private void writeOptimisticUpdatesAndPublish(final InterceptorRequest request) {
    dispatcher.execute(new Runnable() {
      @Override public void run() {
        try {
          if (request.optimisticUpdates.isPresent()) {
            Operation.Data optimisticUpdates = request.optimisticUpdates.get();
            apolloStore.writeOptimisticUpdatesAndPublish(request.operation, optimisticUpdates, request.uniqueId)
                .execute();
          }
        } catch (Exception e) {
          logger.e(e, "failed to write operation optimistic updates, for: %s", request.operation);
        }
      }
    });
  }

  private void rollbackOptimisticUpdatesAndPublish(final InterceptorRequest request) {
    dispatcher.execute(new Runnable() {
      @Override public void run() {
        try {
          apolloStore.rollbackOptimisticUpdatesAndPublish(request.uniqueId).execute();
        } catch (Exception e) {
          logger.e(e, "failed to rollback operation optimistic updates, for: %s", request.operation);
        }
      }
    });
  }

  private Set<String> rollbackOptimisticUpdates(final InterceptorRequest request) {
    try {
      return apolloStore.rollbackOptimisticUpdates(request.uniqueId).execute();
    } catch (Exception e) {
      logger.e(e, "failed to rollback operation optimistic updates, for: %s", request.operation);
      return Collections.emptySet();
    }
  }

  private void publishCacheKeys(final Set<String> cacheKeys) {
    dispatcher.execute(new Runnable() {
      @Override public void run() {
        try {
          apolloStore.publish(cacheKeys);
        } catch (Exception e) {
          logger.e(e, "Failed to publish cache changes");
        }
      }
    });
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloCacheInterceptor.kt
package com.apollographql.apollo.internal.interceptor

import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.internal.ApolloLogger
import com.apollographql.apollo.cache.ApolloCacheHeaders
import com.apollographql.apollo.cache.normalized.ApolloStore
import com.apollographql.apollo.cache.normalized.Record
import com.apollographql.apollo.cache.normalized.internal.Transaction
import com.apollographql.apollo.cache.normalized.internal.WriteableStore
import com.apollographql.apollo.exception.ApolloException
import com.apollographql.apollo.exception.ApolloGenericException
import com.apollographql.apollo.interceptor.ApolloInterceptor
import com.apollographql.apollo.interceptor.ApolloInterceptor.CallBack
import com.apollographql.apollo.interceptor.ApolloInterceptor.FetchSourceType
import com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorRequest
import com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorResponse
import com.apollographql.apollo.interceptor.ApolloInterceptorChain
import java.lang.Runnable
import java.util.ArrayList
import java.util.HashSet
import java.util.concurrent.Executor

/**
 * ApolloCacheInterceptor is a concrete [ApolloInterceptor] responsible for serving requests from the normalized
 * cache if [InterceptorRequest.fetchFromCache] is true. Saves all network responses to cache.
 */
class ApolloCacheInterceptor(
    apolloStore: ApolloStore,
    dispatcher: Executor,
    logger: ApolloLogger, writeToCacheAsynchronously: Boolean) : ApolloInterceptor {
  val apolloStore: ApolloStore
  private val dispatcher: Executor
  private val writeToCacheAsynchronously: Boolean
  val logger: ApolloLogger

  @Volatile
  var disposed = false
  override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                              dispatcher: Executor, callBack: CallBack) {
    dispatcher.execute(Runnable {
      if (disposed) return@Runnable
      if (request.fetchFromCache) {
        callBack.onFetch(FetchSourceType.CACHE)
        val cachedResponse: InterceptorResponse
        try {
          cachedResponse = resolveFromCache(request)
          callBack.onResponse(cachedResponse)
          callBack.onCompleted()
        } catch (e: ApolloException) {
          callBack.onFailure(e)
        }
      } else {
        writeOptimisticUpdatesAndPublish(request)
        chain.proceedAsync(request, dispatcher, object : CallBack {
          override fun onResponse(response: InterceptorResponse) {
            if (disposed) return
            cacheResponseAndPublish(request, response, writeToCacheAsynchronously)
            callBack.onResponse(response)
            callBack.onCompleted()
          }

          override fun onFailure(e: ApolloException) {
            rollbackOptimisticUpdatesAndPublish(request)
            callBack.onFailure(e)
          }

          override fun onCompleted() {
            // call onCompleted in onResponse
          }

          override fun onFetch(sourceType: FetchSourceType) {
            callBack.onFetch(sourceType)
          }
        })
      }
    })
  }

  override fun dispose() {
    disposed = true
  }

  @Throws(ApolloException::class)
  fun resolveFromCache(request: InterceptorRequest): InterceptorResponse {
    val apolloStoreOperation = apolloStore.readOperationInternal(
        request.operation,
        request.cacheHeaders)
    val cachedResponse = apolloStoreOperation.execute()
    if (cachedResponse.data != null) {
      logger.d("Cache HIT for operation %s", request.operation.name())
      return InterceptorResponse(null, cachedResponse, emptySet())
    }
    logger.d("Cache MISS for operation %s", request.operation.name())
    throw ApolloGenericException(String.format("Cache miss for operation %s", request.operation.name()))
  }

  fun cacheResponse(networkResponse: InterceptorResponse,
                    request: InterceptorRequest): Set<String> {
    if (networkResponse.parsedResponse.isPresent
        && networkResponse.parsedResponse.get()!!.hasErrors()
        && !request.cacheHeaders.hasHeader(ApolloCacheHeaders.STORE_PARTIAL_RESPONSES)) {
      return emptySet()
    }
    val records = networkResponse.cacheRecords.orNull()?.let { records ->
      val result: MutableList<Record> = ArrayList(records.size)
      for (record in records) {
        result.add(record.toBuilder().mutationId(request.uniqueId).build())
      }
      result
    }
    return if (records == null) {
      emptySet()
    } else try {
      apolloStore.writeTransaction(object : Transaction<WriteableStore, Set<String>> {
        override fun execute(cache: WriteableStore): Set<String>? {
          return cache.merge(records, request.cacheHeaders)
        }
      })
    } catch (e: Exception) {
      logger.e("Failed to cache operation response", e)
      emptySet()
    }
  }

  fun cacheResponseAndPublish(request: InterceptorRequest, networkResponse: InterceptorResponse, async: Boolean) {
    if (async) {
      dispatcher.execute { cacheResponseAndPublishSynchronously(request, networkResponse) }
    } else {
      cacheResponseAndPublishSynchronously(request, networkResponse)
    }
  }

  fun cacheResponseAndPublishSynchronously(request: InterceptorRequest, networkResponse: InterceptorResponse) {
    try {
      val networkResponseCacheKeys = cacheResponse(networkResponse, request)
      val rolledBackCacheKeys = rollbackOptimisticUpdates(request)
      val changedCacheKeys: MutableSet<String> = HashSet()
      changedCacheKeys.addAll(rolledBackCacheKeys)
      changedCacheKeys.addAll(networkResponseCacheKeys)
      publishCacheKeys(changedCacheKeys)
    } catch (rethrow: Exception) {
      rollbackOptimisticUpdatesAndPublish(request)
      throw rethrow
    }
  }

  fun writeOptimisticUpdatesAndPublish(request: InterceptorRequest) {
    dispatcher.execute {
      try {
        if (request.optimisticUpdates.isPresent) {
          val optimisticUpdates = request.optimisticUpdates.get()
          apolloStore.writeOptimisticUpdatesAndPublish(request.operation as Operation<Operation.Data>, optimisticUpdates, request.uniqueId)
              .execute()
        }
      } catch (e: Exception) {
        logger.e(e, "failed to write operation optimistic updates, for: %s", request.operation)
      }
    }
  }

  fun rollbackOptimisticUpdatesAndPublish(request: InterceptorRequest) {
    dispatcher.execute {
      try {
        apolloStore.rollbackOptimisticUpdatesAndPublish(request.uniqueId).execute()
      } catch (e: Exception) {
        logger.e(e, "failed to rollback operation optimistic updates, for: %s", request.operation)
      }
    }
  }

  fun rollbackOptimisticUpdates(request: InterceptorRequest): Set<String> {
    return try {
      apolloStore.rollbackOptimisticUpdates(request.uniqueId).execute()
    } catch (e: Exception) {
      logger.e(e, "failed to rollback operation optimistic updates, for: %s", request.operation)
      emptySet()
    }
  }

  fun publishCacheKeys(cacheKeys: Set<String>?) {
    dispatcher.execute {
      try {
        apolloStore.publish(cacheKeys!!)
      } catch (e: Exception) {
        logger.e(e, "Failed to publish cache changes")
      }
    }
  }

  init {
    this.apolloStore = apolloStore
    this.dispatcher = dispatcher
    this.logger = logger
    this.writeToCacheAsynchronously = writeToCacheAsynchronously
  }
}
